crates/bitwarden/src/crypto/enc_string/asymmetric.rs

```diff
@@ -0,0 +1,272 @@
+            }
+            #[allow(deprecated)]
+            ("6", 2) => {
+                let data = from_b64_vec(parts[0])?;
+                let mac: Vec<u8> = from_b64_vec(parts[1])?;
+                Ok(AsymmEncString::Rsa2048_OaepSha1_HmacSha256_B64 { data, mac })
+            }
+
+            (enc_type, parts) => Err(EncStringParseError::InvalidType {
+                enc_type: enc_type.to_string(),
+                parts,
+            }
+            .into()),
+        }
+    }
+}
+
+#[allow(unused)]
+impl AsymmEncString {
+    /// TODO: Convert this to a trait method
+    #[cfg(feature = "internal")]
+    pub(crate) fn decrypt_with_private_key(&self, key: &RsaPrivateKey) -> Result<Vec<u8>> {
+        Ok(match self {
+            Self::Rsa2048_OaepSha256_B64 { data } => key.decrypt(Oaep::new::<sha2::Sha256>(), data),
+            Self::Rsa2048_OaepSha1_B64 { data } => key.decrypt(Oaep::new::<sha1::Sha1>(), data),
+            #[allow(deprecated)]
+            Self::Rsa2048_OaepSha256_HmacSha256_B64 { data, mac: _ } => {
+                key.decrypt(Oaep::new::<sha2::Sha256>(), data)
+            }
+            #[allow(deprecated)]
+            Self::Rsa2048_OaepSha1_HmacSha256_B64 { data, mac: _ } => {
+                key.decrypt(Oaep::new::<sha1::Sha1>(), data)
+            }
+        }
+        .map_err(|_| CryptoError::KeyDecrypt)?)
+    }
+}
+
+impl Display for AsymmEncString {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let parts: Vec<&[u8]> = match self {
+            AsymmEncString::Rsa2048_OaepSha256_B64 { data } => vec![data],
+            AsymmEncString::Rsa2048_OaepSha1_B64 { data } => vec![data],
+            #[allow(deprecated)]
+            AsymmEncString::Rsa2048_OaepSha256_HmacSha256_B64 { data, mac } => vec![data, mac],
+            #[allow(deprecated)]
+            AsymmEncString::Rsa2048_OaepSha1_HmacSha256_B64 { data, mac } => vec![data, mac],
+        };
+
+        let encoded_parts: Vec<String> = parts
+            .iter()
+            .map(|part| BASE64_ENGINE.encode(part))
+            .collect();
+
+        write!(f, "{}.{}", self.enc_type(), encoded_parts.join("|"))?;
+
+        Ok(())
+    }
+}
+
+impl<'de> Deserialize<'de> for AsymmEncString {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        struct CSVisitor;
+        impl Visitor<'_> for CSVisitor {
+            type Value = AsymmEncString;
+
+            fn expecting(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+                write!(f, "a valid string")
+            }
+
+            fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+            where
+                E: serde::de::Error,
+            {
+                AsymmEncString::from_str(v).map_err(|e| E::custom(format!("{:?}", e)))
+            }
+        }
+
+        deserializer.deserialize_str(CSVisitor)
+    }
+}
```

Comment:
This looks mostly identical to Symm, is there a way to avoid duplicating as much code? Maybe if we have the generic higher EncString enum and delgate it to that?

Reply:
I've extracted the visitor implementation out, which should help a fair bit